src/bus.rs

```diff
@@ -1,54 +1,165 @@
-use crate::joypad::Joypad;
+use crate::traits::{Device, Memory};
 
 const RAM_MIRRORING_MASK: u16 = 0b0000_0111_1111_1111;
 const PPU_REGISTERS_MIRRORING_MASK: u16 = 0b0010_0000_0000_0111;
 
+enum RunMode {
+    Read,
+    Write,
+}
+
 pub struct Bus {
-    pub(crate) memory:  *mut [u8; 0xFFFF],
-    pub(crate) joypad_1: *mut Joypad,
-    pub(crate) joypad_2: *mut Joypad,
+    memory: *mut [u8; 0xFFFF],
+    devices: Vec<(std::ops::Range<usize>, *mut dyn Device)>,
```

Comment:
is the `dyn` keyword actually needed?

Reply:
https://compiler-explorer.com/z/8h4hKzKaK

im confused, I don't understand why compiler complains in our code

Reply:
ok got it
```
  = note: `#[warn(bare_trait_objects)]` on by default
   = warning: this is accepted in the current edition (Rust 2015) but is a hard error in Rust 2021!
   = note: for more information, see <https://doc.rust-lang.org/nightly/edition-guide/rust-2021/warnings-promoted-to-error.html>
help: use `dyn`
   |
17 -     devices: Vec<*mut Device>,
17 +     devices: Vec<*mut dyn Device>,
```